File: src/services/converters/documentConverter.js
import { createPdfFromText } from '../../utils/pdfGenerator.js';
import { extractTextFromDocx, extractTextFromOdt, extractTextFromEpub } from '../../utils/zipReader.js';

export async function convertDocument(file, targetExt) {
  const cleanTarget = targetExt.toLowerCase();
  const sourceExt = file.name.split('.').pop()?.toLowerCase() || '';

  const contentText = await extractDocumentText(file, sourceExt);

  let outputBlob;
  let mimeType = 'text/plain';

  if (cleanTarget === 'txt') {
    outputBlob = new Blob([contentText], { type: 'text/plain;charset=utf-8' });
    mimeType = 'text/plain';
  } else if (cleanTarget === 'html') {
    const htmlOutput = formatToHTML(contentText, sourceExt, file.name);
    outputBlob = new Blob([htmlOutput], { type: 'text/html;charset=utf-8' });
    mimeType = 'text/html';
  } else if (cleanTarget === 'md') {
    const mdOutput = formatToMarkdown(contentText);
    outputBlob = new Blob([mdOutput], { type: 'text/markdown;charset=utf-8' });
    mimeType = 'text/markdown';
  } else if (cleanTarget === 'rtf') {
    const rtfOutput = formatToRtf(contentText);
    outputBlob = new Blob([rtfOutput], { type: 'application/rtf;charset=utf-8' });
    mimeType = 'application/rtf';
  } else if (cleanTarget === 'pdf') {
    outputBlob = createPdfFromText(file.name, contentText);
    mimeType = 'application/pdf';
  } else if (cleanTarget === 'docx' || cleanTarget === 'doc' || cleanTarget === 'odt' || cleanTarget === 'epub') {
    const wordHtml = formatToWordHTML(contentText, sourceExt, file.name);
    outputBlob = new Blob([wordHtml], { type: 'application/msword;charset=utf-8' });
    mimeType = 'application/msword';
  } else {
    outputBlob = new Blob([contentText], { type: 'text/plain;charset=utf-8' });
  }

  const baseName = file.name.substring(0, file.name.lastIndexOf('.')) || file.name;
  return {
    blob: outputBlob,
    fileName: `${baseName}.${cleanTarget}`,
    mimeType,
  };
}

async function extractDocumentText(file, sourceExt) {
  try {
    const arrayBuffer = await file.arrayBuffer();

    if (sourceExt === 'docx') {
      const docxText = await extractTextFromDocx(arrayBuffer);
      if (docxText && docxText.trim().length > 0) return docxText;
    } else if (sourceExt === 'odt') {
      const odtText = await extractTextFromOdt(arrayBuffer);
      if (odtText && odtText.trim().length > 0) return odtText;
    } else if (sourceExt === 'epub') {
      const epubText = await extractTextFromEpub(arrayBuffer);
      if (epubText && epubText.trim().length > 0) return epubText;
    } else if (sourceExt === 'doc') {
      return parseLegacyDocToText(arrayBuffer, file.name);
    } else if (sourceExt === 'rtf') {
      const rawRtf = new TextDecoder('utf-8').decode(arrayBuffer);
      return parseRtfToText(rawRtf);
    } else if (sourceExt === 'html' || sourceExt === 'htm') {
      const rawHtml = new TextDecoder('utf-8').decode(arrayBuffer);
      return parseHtmlToText(rawHtml);
    } else if (sourceExt === 'pdf') {
      return parsePdfToText(arrayBuffer, file.name);
    }

    // Default for TXT, MD, TEX, and fallback text reading
    const text = new TextDecoder('utf-8').decode(arrayBuffer);
    if (text && !text.includes('\x00')) {
      return stripBinaryJunk(text);
    }

    return parseLegacyDocToText(arrayBuffer, file.name);
  } catch (err) {
    console.warn('Text extraction error:', err);
    return `Document Content from ${file.name}`;
  }
}

function parseLegacyDocToText(arrayBuffer, fileName) {
  try {
    // 1. Try UTF-16LE decoding (standard for OLE2 binary .doc files)
    const utf16Str = new TextDecoder('utf-16le', { fatal: false }).decode(arrayBuffer);
    const utf16Words = utf16Str.match(/[\x20-\x7E]{4,}/g);
    if (utf16Words) {
      const cleanUtf16 = utf16Words.filter(
        w => !w.includes('Root Entry') && !w.includes('WordDocument') && !w.includes('CompObj') && !w.includes('ObjectPool') && w.trim().length > 3
      );
      if (cleanUtf16.length > 0) return cleanUtf16.join('\n');
    }

    // 2. Try UTF-8 / ASCII decoding
    const utf8Str = new TextDecoder('utf-8', { fatal: false }).decode(arrayBuffer);
    const utf8Words = utf8Str.match(/[\x20-\x7E]{4,}/g);
    if (utf8Words) {
      const cleanUtf8 = utf8Words.filter(
        w => !w.includes('Root Entry') && !w.includes('WordDocument') && !w.includes('CompObj') && !w.includes('ObjectPool') && w.trim().length > 3
      );
      if (cleanUtf8.length > 0) return cleanUtf8.join('\n');
    }
  } catch (e) {
    console.warn('parseLegacyDocToText error:', e);
  }

  return `Extracted Document Content from ${fileName}`;
}

function parseRtfToText(rtfStr) {
  return rtfStr
    .replace(/\\fonttbl[\s\S]*?\}/gi, '')
    .replace(/\\colortbl[\s\S]*?\}/gi, '')
    .replace(/\\stylesheet[\s\S]*?\}/gi, '')
    .replace(/\\par/gi, '\n')
    .replace(/\\line/gi, '\n')
    .replace(/\\[a-z]+\d*/gi, '')
    .replace(/[\{\}]/g, '')
    .replace(/\s+/g, ' ')
    .trim();
}

function parseHtmlToText(htmlStr) {
  return htmlStr
    .replace(/<style[^>]*>[\s\S]*?<\/style>/gi, '')
    .replace(/<script[^>]*>[\s\S]*?<\/script>/gi, '')
    .replace(/<br\s*\/?>/gi, '\n')
    .replace(/<\/p>/gi, '\n\n')
    .replace(/<\/h[1-6]>/gi, '\n\n')
    .replace(/<[^>]*>/g, '')
    .replace(/&nbsp;/g, ' ')
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .trim();
}

function parsePdfToText(arrayBuffer, fileName) {
  const decoder = new TextDecoder('utf-8');
  const raw = decoder.decode(arrayBuffer);

  const matches = raw.match(/\((.*?)\)\s*Tj/g);
  if (matches && matches.length > 0) {
    const extracted = matches
      .map(m => m.replace(/^\(/, '').replace(/\)\s*Tj$/, '').replace(/\\/g, ''))
      .filter(t => t.trim().length > 1);
    if (extracted.length > 0) return extracted.join('\n');
  }

  const wordMatches = raw.match(/[\x20-\x7E]{4,}/g);
  if (wordMatches) {
    const cleanWords = wordMatches.filter(
      w => !w.startsWith('/') && !w.startsWith('%') && !w.includes('obj') && !w.includes('endobj') && !w.includes('stream') && !w.includes('xref')
    );
    if (cleanWords.length > 0) return cleanWords.join('\n');
  }

  return `Extracted Document Content from ${fileName}`;
}

function stripBinaryJunk(text) {
  return text
    .replace(/[\x00-\x08\x0B\x0C\x0E-\x1F]/g, '')
    .replace(/\s+/g, ' ')
    .trim();
}

function formatToWordHTML(text, sourceExt, fileName) {
  const paragraphs = text.split('\n').map(p => p.trim()).filter(Boolean);
  const bodyContent = paragraphs.map(p => `<p>${escapeHtml(p)}</p>`).join('\n');

  return `<html xmlns:o='urn:schemas-microsoft-microsoft-com:office:office' xmlns:w='urn:schemas-microsoft-microsoft-com:office:word' xmlns='http://www.w3.org/TR/REC-html40'>
<head>
  <meta charset='utf-8'>
  <title>${escapeHtml(fileName)}</title>
  <!--[if gte mso 9]>
  <xml>
    <w:WordDocument>
      <w:View>Normal</w:View>
      <w:Zoom>100</w:Zoom>
      <w:DoNotOptimizeForBrowser/>
    </w:WordDocument>
  </xml>
  <![endif]-->
  <style>
    body { font-family: 'Calibri', 'Segoe UI', Arial, sans-serif; font-size: 11pt; line-height: 1.6; padding: 24pt; color: #1e1035; }
    h1 { font-size: 18pt; color: #7c3aed; margin-bottom: 12pt; border-bottom: 2px solid #e9d5ff; padding-bottom: 6pt; }
    p { margin-bottom: 10pt; line-height: 1.6; }
  </style>
</head>
<body>
  <h1>${escapeHtml(fileName.replace(/\.[^/.]+$/, ''))}</h1>
  ${bodyContent}
</body>
</html>`;
}

function formatToHTML(text, sourceExt, fileName) {
  const paragraphs = text.split('\n').map(p => p.trim()).filter(Boolean);
  const bodyContent = paragraphs.map(p => `<p>${escapeHtml(p)}</p>`).join('\n');

  return `<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <title>${escapeHtml(fileName)}</title>
  <style>
    body { font-family: 'Segoe UI', Arial, sans-serif; line-height: 1.6; padding: 2.5rem; max-width: 800px; margin: 0 auto; color: #1e1035; background-color: #faf5ff; }
    h1 { color: #7c3aed; border-bottom: 2px solid #e9d5ff; padding-bottom: 0.5rem; }
    p { margin-bottom: 1rem; color: #2e1065; }
  </style>
</head>
<body>
  <h1>${escapeHtml(fileName.replace(/\.[^/.]+$/, ''))}</h1>
  ${bodyContent}
</body>
</html>`;
}

function formatToMarkdown(text) {
  const paragraphs = text.split('\n').map(p => p.trim()).filter(Boolean);
  return paragraphs.join('\n\n');
}

function formatToRtf(text) {
  const clean = text
    .replace(/\\/g, '\\\\')
    .replace(/\{/g, '\\{')
    .replace(/\}/g, '\\}')
    .replace(/\n/g, '\\par\n');

  return `{\\rtf1\\ansi\\deff0{\\fonttbl{\\f0 Segoe UI;}}\\f0\\fs24 ${clean}}`;
}

function escapeHtml(str) {
  return String(str)
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;');
}

File: converters/document_converter.py
import os
import io
import re

def convert_document(input_path_or_bytes, file_name, target_ext):
    clean_target = target_ext.lower().strip('.')
    source_ext = os.path.splitext(file_name)[1].lower().strip('.')

    if isinstance(input_path_or_bytes, bytes):
        input_bytes = input_path_or_bytes
    else:
        with open(input_path_or_bytes, 'rb') as f:
            input_bytes = f.read()

    extracted_text = _extract_text(input_bytes, source_ext, file_name)

    output_bytes = b''
    mime_type = 'text/plain'

    if clean_target == 'txt':
        output_bytes = extracted_text.encode('utf-8')
        mime_type = 'text/plain; charset=utf-8'
    elif clean_target == 'html':
        output_bytes = _format_to_html(extracted_text, file_name).encode('utf-8')
        mime_type = 'text/html; charset=utf-8'
    elif clean_target == 'md':
        output_bytes = _format_to_markdown(extracted_text).encode('utf-8')
        mime_type = 'text/markdown; charset=utf-8'
    elif clean_target == 'rtf':
        output_bytes = _format_to_rtf(extracted_text).encode('utf-8')
        mime_type = 'application/rtf'
    elif clean_target == 'pdf':
        output_bytes = _create_pdf_from_text(file_name, extracted_text)
        mime_type = 'application/pdf'
    elif clean_target in ['docx', 'doc', 'odt', 'epub']:
        output_bytes = _create_docx_from_text(file_name, extracted_text)
        mime_type = 'application/vnd.openxmlformats-officedocument.wordprocessingml.document'
    else:
        output_bytes = extracted_text.encode('utf-8')
        mime_type = 'text/plain; charset=utf-8'

    base_name = os.path.splitext(file_name)[0]
    out_file_name = f"{base_name}.{clean_target}"

    return {
        "bytes": output_bytes,
        "file_name": out_file_name,
        "mime_type": mime_type
    }

def _extract_text(input_bytes, source_ext, file_name):
    if source_ext == 'docx':
        try:
            import docx
            doc = docx.Document(io.BytesIO(input_bytes))
            paragraphs = [p.text.strip() for p in doc.paragraphs if p.text.strip()]
            for table in doc.tables:
                for row in table.rows:
                    row_text = " | ".join([cell.text.strip() for cell in row.cells if cell.text.strip()])
                    if row_text:
                        paragraphs.append(row_text)
            if paragraphs:
                return "\n\n".join(paragraphs)
        except Exception as e:
            pass

    if sourceExt_is_pdf(source_ext, input_bytes):
        try:
            import pypdf
            reader = pypdf.PdfReader(io.BytesIO(input_bytes))
            text_pages = [page.extract_text() for page in reader.pages if page.extract_text()]
            if text_pages:
                return "\n\n".join(text_pages)
        except Exception as e:
            pass

    if source_ext in ['html', 'htm']:
        try:
            from bs4 import BeautifulSoup
            soup = BeautifulSoup(input_bytes, 'html.parser')
            for script in soup(["script", "style"]):
                script.extract()
            text = soup.get_text(separator='\n')
            lines = [line.strip() for line in text.splitlines() if line.strip()]
            if lines:
                return "\n".join(lines)
        except Exception as e:
            pass

    if source_ext == 'rtf':
        try:
            text = input_bytes.decode('utf-8', errors='ignore')
            text = re.sub(r'\\fonttbl[\s\S]*?\}', '', text)
            text = re.sub(r'\\colortbl[\s\S]*?\}', '', text)
            text = re.sub(r'\\[a-z]+\d*', '', text)
            text = re.sub(r'[\{\}]', '', text)
            lines = [line.strip() for line in text.splitlines() if line.strip()]
            if lines:
                return "\n".join(lines)
        except Exception as e:
            pass

    try:
        text = input_bytes.decode('utf-8', errors='ignore')
        text = re.sub(r'[\x00-\x08\x0B\x0C\x0E-\x1F]', '', text)
        lines = [line.strip() for line in text.splitlines() if line.strip()]
        if lines:
            return "\n".join(lines)
    except Exception:
        pass

    return f"Content extracted from {file_name}"

def sourceExt_is_pdf(source_ext, input_bytes):
    return source_ext == 'pdf' or input_bytes.startswith(b'%PDF')

def _create_pdf_from_text(title, text):
    try:
        from reportlab.lib.pagesizes import letter
        from reportlab.pdfgen import canvas

        buffer = io.BytesIO()
        pdf = canvas.Canvas(buffer, pagesize=letter)
        pdf.setTitle(title)

        pdf.setFont("Helvetica-Bold", 16)
        pdf.setFillColorRGB(0.48, 0.22, 0.92) # Purple header
        pdf.drawString(50, 750, title)

        pdf.setFont("Helvetica", 10)
        pdf.setFillColorRGB(0.12, 0.06, 0.21)

        y = 720
        lines = text.split('\n')
        for line in lines:
            if y < 50:
                pdf.showPage()
                pdf.setFont("Helvetica", 10)
                pdf.setFillColorRGB(0.12, 0.06, 0.21)
                y = 750

            wrapped_chunks = [line[i:i+90] for i in range(0, len(line), 90)] or [""]
            for chunk in wrapped_chunks:
                pdf.drawString(50, y, chunk)
                y -= 14
                if y < 50:
                    pdf.showPage()
                    pdf.setFont("Helvetica", 10)
                    pdf.setFillColorRGB(0.12, 0.06, 0.21)
                    y = 750

        pdf.save()
        return buffer.getvalue()
    except Exception as e:
        # Fallback simple PDF generator if reportlab is not installed yet
        return _fallback_pdf(title, text)

def _fallback_pdf(title, text):
    clean_text = text.replace('(', '\\(').replace(')', '\\)')
    lines = clean_text.split('\n')
    stream_content = f"BT /F1 12 Tf 50 750 Td ({title}) Tj 0 -20 Td "
    for line in lines[:50]:
        stream_content += f"({line[:80]}) Tj 0 -14 Td "
    stream_content += "ET"

    pdf_body = f"""%PDF-1.4
1 0 obj <</Type /Catalog /Pages 2 0 R>> endobj
2 0 obj <</Type /Pages /Kids [3 0 R] /Count 1>> endobj
3 0 obj <</Type /Page /Parent 2 0 R /Resources <</Font <</F1 4 0 R>>>> /Contents 5 0 R>> endobj
4 0 obj <</Type /Font /Subtype /Type1 /BaseFont /Helvetica>> endobj
5 0 obj <</Length {len(stream_content)}>>
stream
{stream_content}
endstream
endobj
xref
0 6
0000000000 65535 f 
0000000010 00000 n 
0000000060 00000 n 
0000000117 00000 n 
0000000223 00000 n 
0000000298 00000 n 
trailer <</Size 6 /Root 1 0 R>>
startxref
{350 + len(stream_content)}
%%EOF"""
    return pdf_body.encode('utf-8')

def _create_docx_from_text(title, text):
    try:
        import docx
        doc = docx.Document()
        doc.add_heading(title, level=1)
        for paragraph in text.split('\n'):
            if paragraph.strip():
                doc.add_paragraph(paragraph.strip())
        buffer = io.BytesIO()
        doc.save(buffer)
        return buffer.getvalue()
    except Exception:
        # Simple HTML Word fallback
        html = f"<html><body><h1>{title}</h1>"
        for p in text.split('\n'):
            if p.strip():
                html += f"<p>{p.strip()}</p>"
        html += "</body></html>"
        return html.encode('utf-8')

def _format_to_html(text, title):
    paragraphs = [f"<p>{p.strip()}</p>" for p in text.split('\n') if p.strip()]
    body = "\n".join(paragraphs)
    return f"""<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <title>{title}</title>
    <style>
        body {{ font-family: 'Segoe UI', sans-serif; line-height: 1.6; padding: 2rem; color: #1e1035; background: #faf5ff; }}
        h1 {{ color: #7c3aed; border-bottom: 2px solid #e9d5ff; }}
    </style>
</head>
<body>
    <h1>{title}</h1>
    {body}
</body>
</html>"""

def _format_to_markdown(text):
    return "\n\n".join([p.strip() for p in text.split('\n') if p.strip()])

def _format_to_rtf(text):
    clean = text.replace('\\', '\\\\').replace('{', '\\{').replace('}', '\\}').replace('\n', '\\par\n')
    return f"{{\\rtf1\\ansi\\deff0{{\\fonttbl{{\\f0 Segoe UI;}}}}\\f0\\fs24 {clean}}}"
